### PR TITLE
Check sha256 of local back source cache files and improve `.dirty` handling on backup upload 

### PR DESCRIPTION
**Pull request by AbrilRBS**
Changelog: Fix: Improve handling of `.dirty` download files when uploading backup sources.
Docs: Omit

### CODE REVIEW

File: conans/client/downloaders/caching_file_downloader.py [hunk 1]
```diff
@@ -61,6 +61,7 @@ def _caching_download(self, urls, file_path,
             remove_if_dirty(cached_path)
 
             if os.path.exists(cached_path):
+                self._file_downloader.check_checksum(cached_path, md5, sha1, sha256)
```
Comment (memsharded) [inline conans/client/downloaders/caching_file_downloader.py]: This won't cover all cases of ``.dirty``, only controlled ones. There can be some cases where things will crash and still the ``.dirty`` file will be there.

Comment (AbrilRBS) [reply conans/client/downloaders/caching_file_downloader.py]: Yup! That's why I was not confident in pushing this without being a draft, I still need to check those cases :)

Thanks for the heads up!

File: conans/client/downloaders/download_cache.py [hunk 2]
```diff
@@ -73,22 +73,33 @@ def should_upload_sources(package):
...
 
-        files_to_upload = []
+        path_backups_contents = []
 
+        dirty_ext = ".dirty"
         for path in os.listdir(path_backups):
+            if remove_if_dirty(os.path.join(path_backups, path)):
+                continue
+            if path.endswith(dirty_ext):
+                # TODO: Clear the dirty file marker if it does not have a matching downloaded file
```
Comment (AbrilRBS) [inline conans/client/downloaders/download_cache.py:83]: When we rethink how the `.dirty` files are handled, we won't need to check here nor do anything special

File: conans/test/integration/cache/backup_sources_test.py [hunk 3]
```diff
@@ -677,3 +678,111 @@ def source(self):
...
+            # This check should go away once we refactor dirty handling
+            assert os.path.exists(os.path.join(self.download_cache_folder, "s", f"{sha256}.dirty"))
+
+        if upload:
+            self.client.run("cache backup-upload")
+        else:
+            self.client.run("source .")
+        assert f"{sha256} is dirty, removing it" in self.client.out
+
+    @pytest.mark.skip("Recover when .dirty files are properly handled")
```
Comment (AbrilRBS) [inline conans/test/integration/cache/backup_sources_test.py:752]: To be recovered after https://github.com/conan-io/conan/issues/15612

File: conans/util/files.py [hunk 4]
```diff
@@ -44,6 +44,8 @@ def remove_if_dirty(item):
             else:
                 rmdir(item)
         clean_dirty(item)
+        return True
+    return False
```
Comment (AbrilRBS) [inline conans/util/files.py:48]: Let's the calling method know if a removal happened, useful when skipping dirty files in the backup upload file gathering

**Review by AbrilRBS — COMMENTED** [review]
Some comments

File: conans/util/files.py (hunk 4 above)
Comment (memsharded) [reply conans/util/files.py:48]: Sure, but this is not returning removal, just that the thing was dirty.

**Review by memsharded — APPROVED** [review]

**Pull request merged**